File: 2024/python/05/part1.py
raw = open('input.dat').readlines()
raw = [line.strip() for line in raw]

empty_line_index = raw.index('')
raw1 = raw[:empty_line_index]
raw2 = raw[empty_line_index+1:]
raw2 = [[int(x) for x in line.split(',')] for line in raw2]

def check_valid(num, rest):
    for r in rest:
        if int(r) in rule_dict and num in rule_dict[int(r)]:
            return
    return True

rule_dict = {}
for rule in raw1:
    x, y = map(int, rule.split('|'))
    if x not in rule_dict:
        rule_dict[x] = []
    rule_dict[x].append(y)

sum_middle = 0
for order in raw2:
    for i, num in enumerate(order):
        if not check_valid(num, order[i:]):
            break
    else:
        sum_middle += order[len(order)//2]

print(sum_middle)

File: 2024/go/05/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open("input.dat")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}

	scanner := bufio.NewScanner(file)

	adjMatrix := make(map[int][]int)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			break
		}
		nums := strings.Split(line, "|")
		adjMatrix[toInt(nums[1])] = append(adjMatrix[toInt(nums[1])], toInt(nums[0]))
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	var cnt int
	for scanner.Scan() {
		line := scanner.Text()
		order := strings.Split(line, ",")

		if isValidOrder(order, adjMatrix) {
			mid := int(len(order) / 2)
			cnt += toInt(order[mid])
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	fmt.Println(cnt)
}

func isValidOrder(order []string, adjMatrix map[int][]int) bool {
	for i := 0; i < len(order); i++ {
		for _, dep := range adjMatrix[toInt(order[i])] {
			if contains(order[i+1:], dep) {
				return false
			}
		}
	}
	return true
}

func toInt(s string) int {
	num, err := strconv.Atoi(s)
	if err != nil {
		fmt.Println("Error converting string to int:", err)
		return -1
	}
	return num
}

func contains(s []string, e int) bool {
	for _, a := range s {
		if toInt(a) == e {
			return true
		}
	}
	return false
}
